File: src/main/java/ch/vilki/jfxldap/backend/CmdLine.java
package ch.vilki.jfxldap.backend;

import ch.vilki.secured.SecureString;
import com.unboundid.ldap.sdk.LDAPException;
import com.unboundid.ldap.sdk.Modification;
import com.unboundid.ldap.sdk.ModificationType;
import com.unboundid.ldap.sdk.ModifyRequest;
import net.sourceforge.argparse4j.ArgumentParsers;
import net.sourceforge.argparse4j.inf.ArgumentParser;
import net.sourceforge.argparse4j.inf.Namespace;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CmdLine implements IProgress {

    static Logger logger = LogManager.getLogger(CmdLine.class);
    ArgumentParser _parser = null;
    Config _configuration;
    Namespace _ns = null;

    public void runCmd(String args[]) throws Exception {
        _configuration = new Config();
        _configuration.openConfiguration(Config.getConfigurationFile());
        initParser();
        Namespace ns;

        try {
            _ns = _parser.parseArgs(args);
        } catch (Exception e1) {
            logger.error("Exception in runCMD," + e1.getMessage(), e1);
            _parser.printUsage();
            return;
        }
        String operation = _ns.get("op");
        if (operation.equalsIgnoreCase("export_collection")) {
            exportCollection();
        }
        if (operation.equalsIgnoreCase("get_enviroments")) getEnviroments();
        if (operation.equalsIgnoreCase("import_collection")) importCollection();
        if (operation.equalsIgnoreCase("delete_collection")) deleteCollection();
        if (operation.equalsIgnoreCase("delete_entries")) deleteEntries();
        if (operation.equalsIgnoreCase("delete_attributes")) deleteAttributes();
    }

    private Connection connectToEnviroment(String enviroment) {
        Connection connection = _configuration.getConnection(enviroment);
        SecureString pass = null;
        try {
            pass = _configuration.getConnectionPassword(connection);
        } catch (Exception e) {
            logger.error("Exception in resolving password of the connection " + e.getMessage(), e);
        }
        try {
            if (pass == null) {
                BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
                System.out.println("Enter password for enviroment->" + enviroment);
                pass = new SecureString(br.readLine());
                if (pass.get_value() == null) return null;
            }
            connection.connect(pass);
        } catch (Exception e) {
            logger.error("Error getting password from user", e);
            return null;
        }
        return connection;
    }

    private void exportCollection() {
        logger.info("Calling export collection method");
        String projectFile = _ns.get("projectFile");
        String enviroment = _ns.get("env");
        String exportDirectory = _ns.get("exportDirectory");
        String projectDir = _ns.get("projectDir");
        if (projectFile == null && projectDir == null) {
            System.out.println("ERROR: no project file or directory defined");
            return;
        }
        if (exportDirectory == null && projectDir == null) {
            Path path = Paths.get(projectFile);
            exportDirectory = path.getParent().toAbsolutePath().toString();
        } else if (exportDirectory == null && projectDir != null) {
            exportDirectory = projectDir;
        }
        logger.info("Connecting to enviroment->" + enviroment);
        Connection connection = connectToEnviroment(enviroment);
        if (projectDir != null) {
            logger.info("Exporting from project directory with all project files->" + projectDir);
            Path source = Paths.get(projectDir);
            try {
                final String exportDir = exportDirectory;
                Files.walk(source).filter(Files::isRegularFile).forEach(x -> {
                    if (x.getFileName().toString().toLowerCase().endsWith("project.xlsx")) {
                        CollectionsProject collectionsProject = new CollectionsProject(_configuration, x.getFileName().toAbsolutePath().toString());
                        try {
                            String exportFileName = x.getFileName() + "_export.ldif";
                            collectionsProject.readProject(x.toAbsolutePath().toString());
                            logger.info("Exporting ldif now to ->" + exportFileName);
                            collectionsProject.exportLdif(exportDir, exportFileName, connection, this);
                        } catch (Exception e) {
                            System.out.println("ERROR Exporting" + e.getMessage());
                            e.printStackTrace();
                        }
                    }


                });
                logger.info("All files exported, disconnect now");
                connection.disconect();
            } catch (Exception e) {
                logger.error("ERROR Exporting" + e.getMessage(), e);
            }
        } else {
            logger.info("Exporting from project file specified->" + projectFile);
            try {
                Path path = Paths.get(projectFile);
                CollectionsProject collectionsProject = new CollectionsProject(_configuration, projectFile);
                String exportFileName = path.getFileName() + "_export.ldif";
                collectionsProject.readProject(projectFile);
                collectionsProject.exportLdif(exportDirectory, exportFileName, connection, this);
                connection.disconect();

            } catch (Exception e) {
                System.err.println("Exception occured during export" + e.getMessage());
                return;
            }
        }
    }

    private void deleteCollection() {
        logger.info("delete collection called ");
        String projectFile = _ns.get("projectFile");
        String enviroment = _ns.get("env");
        if (enviroment == null) {
            System.out.println("ERROR: enviroment not  provided");
            return;
        }
        if (projectFile == null) {
            System.out.println("ERROR: no project file  provided");
            return;
        }
        Connection connection = connectToEnviroment(enviroment);
        CollectionsProject collectionsProject = new CollectionsProject(_configuration, projectFile, connection);
        try {
            collectionsProject.readProject(projectFile);
            collectionsProject.deleteAllCollectionEntriesFromEnviromentWithoutParent(this);
        } catch (Exception e) {
            logger.error("Exception during load of project", e);
            return;
        }
    }

    private void importCollection() {
        logger.info("Import collection called ");
        String importOptions = _ns.get("importOptions");
        if (importOptions == null) {
            logger.error("No import options have been defined");
            _parser.printHelp();
            return;
        }
        String inputFile = _ns.get("inputFile");
        String importDirectory = _ns.get("importDirectory");
        if (inputFile == null && importDirectory == null) {
            logger.error("No input file or directory have been provided");
            _parser.printUsage();
            return;
        }
        String enviroment = _ns.get("env");
        if (enviroment == null) {
            logger.error("No enviroment to import to has been provided");
            _parser.printUsage();
            return;
        }
        Connection connection = connectToEnviroment(enviroment);
        if (connection == null) {
            System.err.println("could not connect to->" + enviroment);
            return;
        }
        CollectionImport collectionImport = new CollectionImport();
        collectionImport.set_progress(this);
        if (importDirectory != null) {
            if (!confirm("Start dir import ->" + importDirectory + " in ->" + connection.getName())) return;
            collectionImport.uploadDirectory(importDirectory, false, connection, null,
                    this, CollectionImport.get_IMPORT_OPTIONS(importOptions));
        } else {
            if (!confirm("Start ldif import ->" + inputFile + " in ->" + connection.getName())) return;
            collectionImport.loadFile(inputFile, null);
            try {
                collectionImport.importInEnviroment(false, connection, CollectionImport.get_IMPORT_OPTIONS(importOptions));
            } catch (Exception e) {
                System.err.println("Exception during import" + e.getMessage());
                return;
            }
        }
    }

    private void deleteEntries() {
        String inputFile = _ns.get("inputFile");
        String enviroment = _ns.get("env");
        if (inputFile == null) {
            System.err.println("No collection project file has been specified, " +
                    "need collection project file location with -projectFile PROJECT_FILE");
            return;
        }
        if (enviroment == null) {
            System.err.println("Enviroment from which entries are to be deleted not specified, specify with -env ENVIROMENT");
            return;
        }
        if (!confirm("WARNING!!! THIS WILL DELETE ENTRIES FROM ->" + enviroment)) return;
        Connection connection = connectToEnviroment(enviroment);
        if (connection == null) {
            System.err.println("Connection not established, break the operation");
            return;
        }
        CollectionImport collectionImport = new CollectionImport();
        collectionImport.set_progress(this);
        try {
            collectionImport.loadFile(inputFile, null);

        } catch (Exception e) {
            System.err.println("Error during reading of project" + e.getMessage());
        }
        collectionImport.deleteEntries(connection, null);
    }

    private void deleteAttributes() throws Exception {
        String inputFile = _ns.get("inputFile");
        String attributes = _ns.get("attributes");
        if (inputFile == null) throw new Exception("No input file has been specified");
        if (!Files.exists(Paths.get(inputFile))) throw new Exception("File does not exist");
        if (attributes == null) throw new Exception("Attributes not specified");
        Connection connection = new Connection(inputFile,this,true);
        connection.set_readOnly(false);
        String split[] = attributes.split(",");
        List<ModifyRequest> modifyRequests = new ArrayList<>();
        for (String dn : connection.get_fileEntries().keySet()) {
            List<Modification> modifications = new ArrayList<>();
            for (String attName : split) {
                Modification modification = new Modification(ModificationType.DELETE, attName);
                modifications.add(modification);
            }

            ModifyRequest modifyRequest = new ModifyRequest(dn, modifications);
            modifyRequests.add(modifyRequest);
        }
        connection.modify(modifyRequests);
    }

    private void getEnviroments() throws LDAPException {

        if (_configuration == null) {
            System.err.println("ERROR: Did not find configuration file, can not proceed");
            return;
        }
        for (String envKey : _configuration._connections.keySet()) {
            System.out.println(envKey);
            Connection connection = _configuration._connections.get(envKey);
            System.out.println(connection.getServer());
            System.out.println(connection.getUser());
            System.out.println(connection.getBaseDN());
            System.out.println("--------------------------------------");
        }
    }

    private boolean confirm(String message) {
        java.util.Scanner s = new java.util.Scanner(System.in);
        System.out.println(message);
        System.out.println("Confirm YES[y] or any other key for NO");
        String name = s.nextLine();
        if (name.equalsIgnoreCase("Y")) return true;
        return false;
    }

    private void initParser() {
        _parser = ArgumentParsers.newArgumentParser("cncOperations")
                .description("Calls different operations")
                .defaultHelp(true);

        _parser.addArgument("-op")
                .choices("export_collection", "get_enviroments", "import_collection", "delete_entries", "delete_collection", "delete_attributes")
                .help("Specify operation type")
                .required(true)
                .dest("op");

        _parser.addArgument("-projectFile")
                .help("Specify project file ")
                .dest("projectFile");

        _parser.addArgument("-projectDir")
                .help("Specify project dir ")
                .dest("projectDir");

        _parser.addArgument("-inputFile")
                .help("Specify input file ")
                .dest("inputFile");

        _parser.addArgument("-importDirectory")
                .help("Specify project dir ")
                .dest("importDirectory");


        _parser.addArgument("-exportFile")
                .help("Specify exportFile file ")
                .dest("exportFile");

        _parser.addArgument("-exportDirectory")
                .help("Specify exportDirectory  ")
                .dest("exportDirectory");

        _parser.addArgument("-env")
                .choices(_configuration._connections.keySet())
                .help("Specify Enviroment " + _configuration._connections.keySet())
                .dest("env");
        _parser.addArgument("-attributes").dest("attributes");


        List<String> importOptions = new ArrayList<>();
        Arrays.stream(CollectionImport.IMPORT_OPTIONS.values()).forEach(x -> importOptions.add(x.toString()));
        _parser.addArgument("-importOptions")
                .choices(importOptions)
                .help("Specify import option type->" + importOptions)
                .dest("importOptions");
    }

    @Override
    public void setProgress(double progress, String description) {
        if (description != null) {
            System.out.println(description + " [" + progress + "]");
        } else {
            System.out.println("Progress" + " [" + progress + "]");
        }
    }

    @Override
    public void signalTaskDone(String taskName, String description, Exception e) {
        if (e != null) {
            System.err.println("ERROR Task->" + taskName + " , description->" + description + "  " + e.toString());
            e.printStackTrace();
        } else {
            System.out.println("TASK DONE->" + taskName + ", description->" + description);
        }

    }

    @Override
    public void setProgress(String taskName, double progress) {
        if (taskName != null) {
            System.out.println(taskName + " [" + progress + "]");
        } else {
            System.out.println("Progress" + " [" + progress + "]");
        }
    }
}

File: src/main/kotlin/ch/vilki/jfxldap/backend/CmdLine.kt
package ch.vilki.jfxldap.backend

import ch.vilki.secured.ApplicationVersion
import ch.vilki.secured.SecureString
import com.unboundid.ldap.sdk.LDAPException
import com.unboundid.ldap.sdk.Modification
import com.unboundid.ldap.sdk.ModificationType
import com.unboundid.ldap.sdk.ModifyRequest
import net.sourceforge.argparse4j.ArgumentParsers
import net.sourceforge.argparse4j.inf.ArgumentParser
import net.sourceforge.argparse4j.inf.Namespace
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import java.io.BufferedReader
import java.io.InputStreamReader
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*

class CmdLine : IProgress {
    companion object {
        private val logger: Logger = LogManager.getLogger(CmdLine::class.java)
    }

    private var _parser: ArgumentParser? = null
    private lateinit var _configuration: Config
    private var _ns: Namespace? = null

    @Throws(Exception::class)
    fun runCmd(args: Array<String>) {

        _configuration = Config()
        _configuration.openConfiguration(Config.configurationFile)
        println("jfxldap ${ApplicationVersion.VERSION}")
        initParser()

        try {
            _ns = _parser?.parseArgs(args)
        } catch (e1: Exception) {
            println(e1.localizedMessage)
            _parser?.printUsage()
            return
        }

        val operation = _ns?.getString("op") ?: return
        when {
            operation.equals("export_collection", ignoreCase = true) -> exportCollection()
            operation.equals("get_enviroments", ignoreCase = true) -> getEnviroments()
            operation.equals("import_collection", ignoreCase = true) -> importCollection()
            operation.equals("delete_collection", ignoreCase = true) -> deleteCollection()
            operation.equals("delete_entries", ignoreCase = true) -> deleteEntries()
            operation.equals("delete_attributes", ignoreCase = true) -> deleteAttributes()
        }
    }

    private fun connectToEnviroment(enviroment: String): Connection? {
        val connection: Connection = _configuration.getConnection(enviroment) ?: return null
        var pass: SecureString? = null
        try {
            pass = _configuration.getConnectionPassword(connection)
        } catch (e: Exception) {
            logger.error("Exception in resolving password of the connection ${e.message}", e)
        }
        
        try {
            if (pass == null) {
                val br = BufferedReader(InputStreamReader(System.`in`))
                println("Enter password for enviroment->$enviroment")
                pass = SecureString(br.readLine())
                if (pass.get_value() == null) return null
            }
            connection.connect(pass)
        } catch (e: Exception) {
            logger.error("Error getting password from user", e)
            return null
        }
        
        return connection
    }

    private fun exportCollection() {
        logger.info("Calling export collection method")
        val projectFile = _ns?.getString("projectFile")
        val enviroment = _ns?.getString("env")
        var exportDirectory = _ns?.getString("exportDirectory")
        val projectDir = _ns?.getString("projectDir")
        
        if (projectFile == null && projectDir == null) {
            println("ERROR: no project file or directory defined")
            return
        }
        
        if (exportDirectory == null && projectDir == null) {
            val path = Paths.get(projectFile)
            exportDirectory = path.parent.toAbsolutePath().toString()
        } else if (exportDirectory == null && projectDir != null) {
            exportDirectory = projectDir
        }
        
        logger.info("Connecting to enviroment->$enviroment")
        val connection = connectToEnviroment(enviroment!!) ?: return
        
        if (projectDir != null) {
            logger.info("Exporting from project directory with all project files->$projectDir")
            val source = Paths.get(projectDir)
            
            try {
                val exportDir = exportDirectory!!
                Files.walk(source).filter(Files::isRegularFile).forEach { x ->
                    if (x.fileName.toString().lowercase().endsWith("project.xlsx")) {
                        val collectionsProject = CollectionsProject(_configuration, x.fileName.toAbsolutePath().toString())
                        try {
                            val exportFileName = "${x.fileName}_export.ldif"
                            collectionsProject.readProject(x.toAbsolutePath().toString())
                            logger.info("Exporting ldif now to ->$exportFileName")
                            collectionsProject.exportLdif(exportDir, exportFileName, connection, this)
                        } catch (e: Exception) {
                            println("ERROR Exporting${e.message}")
                            e.printStackTrace()
                        }
                    }
                }
                
                logger.info("All files exported, disconnect now")
                connection.disconect()
            } catch (e: Exception) {
                logger.error("ERROR Exporting${e.message}", e)
            }
        } else {
            logger.info("Exporting from project file specified->$projectFile")
            
            try {
                val path = Paths.get(projectFile!!)
                val collectionsProject = CollectionsProject(_configuration, projectFile)
                val exportFileName = "${path.fileName}_export.ldif"
                collectionsProject.readProject(projectFile)
                collectionsProject.exportLdif(exportDirectory!!, exportFileName, connection, this)
                connection.disconect()
            } catch (e: Exception) {
                System.err.println("Exception occured during export${e.message}")
                return
            }
        }
    }

    private fun deleteCollection() {
        logger.info("delete collection called ")
        val projectFile = _ns?.getString("projectFile")
        val enviroment = _ns?.getString("env")
        
        if (enviroment == null) {
            println("ERROR: enviroment not provided")
            return
        }
        
        if (projectFile == null) {
            println("ERROR: no project file provided")
            return
        }
        
        val connection = connectToEnviroment(enviroment)
        val collectionsProject = CollectionsProject(_configuration, projectFile, connection)
        
        try {
            collectionsProject.readProject(projectFile)
            collectionsProject.deleteAllCollectionEntriesFromEnviromentWithoutParent(this)
        } catch (e: Exception) {
            logger.error("Exception during load of project", e)
            return
        }
    }

    private fun importCollection() {
        logger.info("Import collection called ")
        val importOptions = _ns?.getString("importOptions")
        
        if (importOptions == null) {
            logger.error("No import options have been defined")
            _parser?.printHelp()
            return
        }
        
        val inputFile = _ns?.getString("inputFile")
        val importDirectory = _ns?.getString("importDirectory")
        
        if (inputFile == null && importDirectory == null) {
            logger.error("No input file or directory have been provided")
            _parser?.printUsage()
            return
        }
        
        val enviroment = _ns?.getString("env")
        
        if (enviroment == null) {
            logger.error("No enviroment to import to has been provided")
            _parser?.printUsage()
            return
        }
        
        val connection = connectToEnviroment(enviroment)
        
        if (connection == null) {
            System.err.println("could not connect to->$enviroment")
            return
        }
        
        val collectionImport = CollectionImport()
        collectionImport.set_progress(this)
        
        if (importDirectory != null) {
            if (!confirm("Start dir import ->$importDirectory in ->${connection.getName()}")) return
            collectionImport.uploadDirectory(
                importDirectory, 
                false, 
                connection, 
                null,
                this, 
                CollectionImport.get_IMPORT_OPTIONS(importOptions)
            )
        } else {
            if (!confirm("Start ldif import ->$inputFile in ->${connection.getName()}")) return
            collectionImport.loadFile(inputFile, null)
            
            try {
                collectionImport.importInEnviroment(false, connection, CollectionImport.get_IMPORT_OPTIONS(importOptions))
            } catch (e: Exception) {
                System.err.println("Exception during import${e.message}")
                return
            }
        }
    }

    private fun deleteEntries() {
        val inputFile = _ns?.getString("inputFile")
        val enviroment = _ns?.getString("env")
        
        if (inputFile == null) {
            System.err.println("No collection project file has been specified, " +
                    "need collection project file location with -projectFile PROJECT_FILE")
            return
        }
        
        if (enviroment == null) {
            System.err.println("Enviroment from which entries are to be deleted not specified, specify with -env ENVIROMENT")
            return
        }
        
        if (!confirm("WARNING!!! THIS WILL DELETE ENTRIES FROM ->$enviroment")) return
        
        val connection = connectToEnviroment(enviroment)
        
        if (connection == null) {
            System.err.println("Connection not established, break the operation")
            return
        }
        
        val collectionImport = CollectionImport()
        collectionImport.set_progress(this)
        
        try {
            collectionImport.loadFile(inputFile, null)
        } catch (e: Exception) {
            System.err.println("Error during reading of project${e.message}")
        }
        
        collectionImport.deleteEntries(connection, null)
    }

    @Throws(Exception::class)
    private fun deleteAttributes() {
        val inputFile = _ns?.getString("inputFile")
        val attributes = _ns?.getString("attributes")
        
        if (inputFile == null) throw Exception("No input file has been specified")
        if (!Files.exists(Paths.get(inputFile))) throw Exception("File does not exist")
        if (attributes == null) throw Exception("Attributes not specified")
        
        val connection = Connection(inputFile, this, true)
        connection.set_readOnly(false)
        
        val split = attributes.split(",").toTypedArray()
        val modifyRequests = ArrayList<ModifyRequest>()
        
        for (dn in connection.get_fileEntries().keys) {
            val modifications = ArrayList<Modification>()
            
            for (attName in split) {
                val modification = Modification(ModificationType.DELETE, attName)
                modifications.add(modification)
            }
            
            val modifyRequest = ModifyRequest(dn, modifications)
            modifyRequests.add(modifyRequest)
        }
        
        connection.modify(modifyRequests)
    }

    @Throws(LDAPException::class)
    private fun getEnviroments() {
        if (_configuration == null) {
            System.err.println("ERROR: Did not find configuration file, can not proceed")
            return
        }
        
        for (envKey in _configuration._connections.keys) {
            println(envKey)
            val connection = _configuration._connections[envKey]
            println(connection?.getServer())
            println(connection?.getUser())
            println(connection?.getBaseDN())
            println("--------------------------------------")
        }
    }

    private fun confirm(message: String): Boolean {
        val s = Scanner(System.`in`)
        println(message)
        println("Confirm YES[y] or any other key for NO")
        val name = s.nextLine()
        return name.equals("Y", ignoreCase = true)
    }

    private fun initParser() {
        _parser = ArgumentParsers.newArgumentParser("cncOperations")
            .description("Calls different operations")
            .defaultHelp(true)

        _parser?.addArgument("-op")
            ?.choices("export_collection", "get_enviroments", "import_collection", "delete_entries", "delete_collection", "delete_attributes")
            ?.help("Specify operation type")
            ?.required(true)
            ?.dest("op")

        _parser?.addArgument("-projectFile")
            ?.help("Specify project file ")
            ?.dest("projectFile")

        _parser?.addArgument("-projectDir")
            ?.help("Specify project dir ")
            ?.dest("projectDir")

        _parser?.addArgument("-inputFile")
            ?.help("Specify input file ")
            ?.dest("inputFile")

        _parser?.addArgument("-importDirectory")
            ?.help("Specify project dir ")
            ?.dest("importDirectory")

        _parser?.addArgument("-exportFile")
            ?.help("Specify exportFile file ")
            ?.dest("exportFile")

        _parser?.addArgument("-exportDirectory")
            ?.help("Specify exportDirectory  ")
            ?.dest("exportDirectory")

        _parser?.addArgument("-env")
            ?.choices(_configuration._connections.keys)
            ?.help("Specify Enviroment ${_configuration._connections.keys}")
            ?.dest("env")
            
        _parser?.addArgument("-attributes")?.dest("attributes")

        val importOptions = ArrayList<String>()
        CollectionImport.IMPORT_OPTIONS.values().forEach { x -> importOptions.add(x.toString()) }
        
        _parser?.addArgument("-importOptions")
            ?.choices(importOptions)
            ?.help("Specify import option type->$importOptions")
            ?.dest("importOptions")
    }

    override fun setProgress(progress: Double, description: String?) {
        if (description != null) {
            println("$description [$progress]")
        } else {
            println("Progress [$progress]")
        }
    }

    override fun signalTaskDone(taskName: String?, description: String?, e: Exception?) {
        if (e != null) {
            System.err.println("ERROR Task->$taskName , description->$description  ${e}")
            e.printStackTrace()
        } else {
            println("TASK DONE->$taskName, description->$description")
        }
    }

    override fun setProgress(taskName: String?, progress: Double) {
        if (taskName != null) {
            println("$taskName [$progress]")
        } else {
            println("Progress [$progress]")
        }
    }
}
